apps/dashboard/src/contract-ui/tabs/tokens/components/airdrop-upload-erc20.tsx

```diff
@@ -0,0 +1,469 @@
+  CircleAlert,
+  Upload,
+} from "lucide-react";
+import Papa from "papaparse";
+import { useCallback, useMemo, useRef, useState } from "react";
+import { type DropzoneOptions, useDropzone } from "react-dropzone";
+import { type Column, usePagination, useTable } from "react-table";
+import { ZERO_ADDRESS } from "thirdweb";
+import { resolveAddress } from "thirdweb/extensions/ens";
+import { isAddress } from "thirdweb/utils";
+import { Button, Heading, Text } from "tw-components";
+import { csvMimeTypes } from "utils/batch";
+
+export interface ERC20AirdropAddressInput {
+  address: string;
+  quantity: string;
+  isValid?: boolean;
+}
+interface AirdropUploadProps {
+  setAirdrop: (airdrop: ERC20AirdropAddressInput[]) => void;
+  onClose: () => void;
+}
+
+export const AirdropUploadERC20: React.FC<AirdropUploadProps> = ({
+  setAirdrop,
+  onClose,
+}) => {
+  const thirdwebClient = useThirdwebClient();
+  const [validAirdrop, setValidAirdrop] = useState<ERC20AirdropAddressInput[]>(
+    [],
+  );
+  const [noCsv, setNoCsv] = useState(false);
+
+  const reset = useCallback(() => {
+    setValidAirdrop([]);
+    setNoCsv(false);
+  }, []);
+
+  const onDrop = useCallback<Required<DropzoneOptions>["onDrop"]>(
+    (acceptedFiles) => {
+      setNoCsv(false);
+
+      const csv = acceptedFiles.find(
+        (f) => csvMimeTypes.includes(f.type) || f.name?.endsWith(".csv"),
+      );
+      if (!csv) {
+        console.error(
+          "No valid CSV file found, make sure you have an address column.",
+        );
+        setNoCsv(true);
+        return;
+      }
+
+      Papa.parse(csv, {
+        header: true,
+        complete: (results) => {
+          const data: ERC20AirdropAddressInput[] = (
+            results.data as ERC20AirdropAddressInput[]
+          )
+            .map(({ address, quantity }) => ({
+              address: (address || "").trim(),
+              quantity: (quantity || "1").trim(),
+            }))
+            .filter(({ address }) => address !== "");
+
+          if (!data[0]?.address) {
+            setNoCsv(true);
+            return;
+          }
+
+          setValidAirdrop(data);
+        },
+      });
+    },
+    [],
+  );
+
+  async function _checkIsAddress(item: ERC20AirdropAddressInput) {
+    const { address, ...rest } = item;
+    // Sending tokens to the zero address will result in an error
+    // if user wants to burn, they should use the Burn tab
+    if (address === ZERO_ADDRESS) {
+      return {
+        address,
+        resolvedAddress: address,
+        isValid: false,
+        ...rest,
+      };
+    }
+
+    let isValid = true;
+    let resolvedAddress = address;
+
+    try {
+      resolvedAddress = isAddress(address)
+        ? address
+        : await resolveAddress({ name: address, client: thirdwebClient });
+      isValid = !!resolvedAddress;
+    } catch {
+      isValid = false;
+    }
+
+    return {
+      address,
+      resolvedAddress,
+      isValid,
+      ...rest,
+    };
+  }
```

Comment:
move this function out of the component, right now every time the component re-renders u are re-creating this function, too